### PR TITLE
feature: Load workspace build logs from streaming

### LINKED ISSUE #1769: Allow build logs to add event in real time instead of needing a page refresh
As a user, I would like to watch the log event populate in real time in the UI when a workspace is building. 

-----
The workspace build logs are not currently updating in real time, they require a page refresh to see events that are occurring after the each page load. 

Steps to repro: 

1. Create a new workspace. 
2. Click into a status that appears in the Timeline section of the Workspaces page to open the logs. 
3. Hand and watch for a few seconds or more to verify you are not seeing live updates of events 
4. Refresh the page to see it update the events.

### LINKED ISSUE #1996: Use stream to load the workspace build logs

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
Closes #1769

### CODE REVIEW

**Review by Kira-Pilot — APPROVED** [review]
Awesome!

File: site/src/xServices/workspaceBuild/workspaceBuildXService.ts [hunk 1]
```diff
@@ -87,16 +109,32 @@ export const workspaceBuildMachine = createMachine(
...
     services: {
       getWorkspaceBuild: (ctx) => API.getWorkspaceBuild(ctx.buildId),
-      getWorkspaceBuildLogs: (ctx) => API.getWorkspaceBuildLogs(ctx.buildId),
+      getLogs: async (ctx) => API.getWorkspaceBuildLogs(ctx.buildId),
+      streamWorkspaceBuildLogs: (ctx) => async (callback) => {
+        const reader = await API.streamWorkspaceBuildLogs(ctx.buildId)
+
+        // Watching for the stream
+        // eslint-disable-next-line no-constant-condition, @typescript-eslint/no-unnecessary-condition
+        while (true) {
```
[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply site/src/xServices/workspaceBuild/workspaceBuildXService.ts:127]: It is a good point. I think for now we can rely on done being always called, why:
- We are using a polyfill(https://canjs.com/doc/can-ndjson-stream.html)  because Firefox does not support some stream APIs
-  I took a look and I didn't find anything related to "read failure" or something related
- If an error is thrown, it will stop the execution anyway
- This error, in particular, looks tricky to handle so I would expect someone having issues with that to think of a proper solution
- In summary, I don't know how to do that and it looks complex to handle, so I would not invest too much time on that right now, but if you have an idea how to solve that, I would love to learn and implement it for sure.

File: site/src/components/WorkspaceBuildLogs/WorkspaceBuildLogs.tsx [hunk 2]
```diff
@@ -35,29 +36,34 @@ const getStageDurationInSeconds = (logs: ProvisionerJobLog[]) => {
...
+        const shouldDisplaySpinner = isWaitingForLogs && isLastStage
+        const shouldDisplayDuration = !isWaitingForLogs && duration
 
         return (
           <div key={stage}>
             <div className={styles.header}>
               <div>{stage}</div>
-              {duration && <div className={styles.duration}>{duration} seconds</div>}
+              {shouldDisplaySpinner && <CircularProgress size={14} className={styles.spinner} />}
+              {shouldDisplayDuration && <div className={styles.duration}>{duration} seconds</div>}
```
[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply site/src/components/WorkspaceBuildLogs/WorkspaceBuildLogs.tsx:66]: Definetely!

File: site/src/xServices/workspaceBuild/workspaceBuildXService.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kira-Pilot) [reply site/src/xServices/workspaceBuild/workspaceBuildXService.ts:127]: Ha! Thanks for the breakdown - what you said makes sense and I think you're right to hold off. If this becomes an issue in the future, we can look into [calling `cancel` on the ReadableStream object](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/cancel).

[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply site/src/xServices/workspaceBuild/workspaceBuildXService.ts:127]: Ahh, nice!
